src/main/java/com/google/devtools/build/remote/client/AbstractRemoteActionCache.java

```diff
@@ -15,17 +15,183 @@
+import com.google.devtools.remoteexecution.v1test.Directory;
+import com.google.devtools.remoteexecution.v1test.DirectoryNode;
+import com.google.devtools.remoteexecution.v1test.FileNode;
+import com.google.devtools.remoteexecution.v1test.Tree;
+import com.google.protobuf.ByteString;
 import java.io.IOException;
 import java.io.OutputStream;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.attribute.PosixFilePermission;
+import java.util.ArrayList;
+import java.util.EnumSet;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import javax.annotation.Nullable;
 
 /** A cache for storing artifacts (input and output) as well as the output of running an action. */
 public abstract class AbstractRemoteActionCache {
+
+  public DigestUtil getDigestUtil() {
+    return digestUtil;
+  }
+
   protected final DigestUtil digestUtil;
 
   public AbstractRemoteActionCache(DigestUtil digestUtil) {
     this.digestUtil = digestUtil;
   }
 
-  /** Downloads a single blob with the specified digest into an output stream. */
+  /**
+   * Download a tree with the {@link Directory} given by digest as the root directory of the tree.
+   *
+   * @param rootDigest The digest of the root {@link Directory} of the tree
+   * @return A tree with the given directory as the root.
+   * @throws IOException in the case that retrieving the blobs required to reconstruct the tree
+   *     failed.
+   */
+  public Tree getTree(Digest rootDigest) throws IOException {
+    byte directoryBytes[];
+    Directory dir;
+    directoryBytes = downloadBlob(rootDigest);
+    dir = Directory.parseFrom(directoryBytes);
+    return getTree(dir);
+  }
+
+  // A default implementation for recursively downloading the tree associated with a directory.
+  // Note: This implementation requires a round trip for each directory in the directory tree.
+  protected Tree getTree(Directory rootDir) throws IOException {
+    List<Directory> children = new ArrayList<>();
+    addChildDirectories(children, rootDir);
+    return Tree.newBuilder().setRoot(rootDir).addAllChildren(children).build();
+  }
+
+  // Recursively add all child directories to the given list.
+  private void addChildDirectories(List<Directory> directories, Directory dir) throws IOException {
+    for (DirectoryNode childNode : dir.getDirectoriesList()) {
+      byte[] childDirBytes = downloadBlob(childNode.getDigest());
+      Directory childDir = Directory.parseFrom(childDirBytes);
+      directories.add(childDir);
+      addChildDirectories(directories, childDir);
+    }
+  }
+
+  /**
+   * Download the full contents of a Directory to a local path given its digest.
+   *
+   * @param downloadPath The path to download the directory contents to.
+   * @param directoryDigest The digest of the Directory to download.
+   * @throws IOException
+   */
+  public void downloadDirectory(Path downloadPath, Digest directoryDigest) throws IOException {
```

Comment:
Nit: here and elsewhere, could you put the path (relating to output) after the directory (relating to input)?

Reply:
Some of these method signatures are copied from their counterparts from the Bazel library, which has them in this order (e.g. https://github.com/bazelbuild/bazel/blob/master/src/main/java/com/google/devtools/build/lib/remote/AbstractRemoteActionCache.java#L218 has output Path as the first parameter), but I think I will change them anyways.

Reply:
Makes sense. In that case, please don't change them :-)
Maintaining consistency is better.